File: ios/App/App/Features/RichChat/RichPendingSteer.swift
import Foundation

enum RichPromptSegment: Sendable, Equatable {
  case text(content: String)
  case file(path: String)
  case attachment(path: String, mimeType: String?)
  case diffComment(path: String, lineNumber: Int64, side: RichDiffSide, staged: Bool, body: String)
  case skill(
    name: String,
    path: String,
    invocation: String,
    provider: String,
    scope: String,
    pluginID: String?,
    pluginName: String?
  )
  case mcp(id: String, name: String)
}

struct RichPendingSteer: Sendable, Equatable, Identifiable {
  let id: String
  let prompt: String
  let segments: [RichPromptSegment]?
  let stagedAtMilliseconds: Double
}

struct RichPendingSteerEnvelope: Sendable, Equatable {
  let threadID: String
  let pending: RichPendingSteer?
}

struct RichSetPendingSteerInput: Sendable, Equatable {
  let prompt: String
  let segments: [RichPromptSegment]?
  let config: [String: RichJSON]
}

struct RichPendingSteerState: Sendable, Equatable {
  let threadID: String
  private(set) var pending: RichPendingSteer?

  init(threadID: String, pending: RichPendingSteer? = nil) {
    self.threadID = threadID
    self.pending = pending
  }

  mutating func apply(_ envelope: RichPendingSteerEnvelope) {
    guard envelope.threadID == threadID else { return }
    pending = envelope.pending
  }
}

enum RichPendingSteerDecoder {
  static func decodeSetBody(_ value: RichJSON) throws -> RichSetPendingSteerInput {
    guard let object = value.objectValue,
      let prompt = RichDecoding.requiredString(object, "prompt", allowEmpty: false),
      let config = object["config"]?.objectValue
    else { throw RichDomainDecodeError.invalidPendingSteer }
    return RichSetPendingSteerInput(
      prompt: prompt,
      segments: try decodeOptionalSegments(object),
      config: config
    )
  }

  static func isValidClearBody(_ value: RichJSON) -> Bool {
    value.objectValue?.isEmpty == true
  }

  static func decodeEnvelope(_ value: RichJSON) throws -> RichPendingSteerEnvelope {
    guard let object = value.objectValue,
      RichDecoding.requiredString(object, "type") == "thread-pending-steer",
      let threadID = RichDecoding.requiredString(object, "threadId", allowEmpty: false),
      let rawPending = object["pending"]
    else { throw RichDomainDecodeError.invalidPendingSteer }
    let pending = rawPending == .null ? nil : try decodePending(rawPending)
    return RichPendingSteerEnvelope(threadID: threadID, pending: pending)
  }

  static func decodeSegment(_ value: RichJSON) throws -> RichPromptSegment {
    guard let object = value.objectValue,
      let kind = RichDecoding.requiredString(object, "kind")
    else { throw RichDomainDecodeError.invalidPendingSteer }
    switch kind {
    case "text":
      guard let content = RichDecoding.requiredString(object, "content") else { break }
      return .text(content: content)
    case "file":
      guard let path = RichDecoding.requiredString(object, "path") else { break }
      return .file(path: path)
    case "attachment":
      guard let path = RichDecoding.requiredString(object, "path") else { break }
      let mime = RichDecoding.optionalString(object, "mimeType")
      guard mime != .invalid else { break }
      return .attachment(path: path, mimeType: mime.value)
    case "diff_comment":
      guard let path = RichDecoding.requiredString(object, "path", allowEmpty: false),
        let line = object["lineNumber"]?.exactInt64Value, line > 0,
        let sideText = RichDecoding.requiredString(object, "side"),
        let side = RichDiffSide(rawValue: sideText),
        let staged = object["staged"]?.boolValue,
        let body = RichDecoding.requiredString(object, "body", allowEmpty: false)
      else { break }
      return .diffComment(path: path, lineNumber: line, side: side, staged: staged, body: body)
    case "skill":
      guard let name = RichDecoding.requiredString(object, "name", allowEmpty: false),
        let path = RichDecoding.requiredString(object, "path", allowEmpty: false),
        let invocation = RichDecoding.requiredString(object, "invocation", allowEmpty: false),
        let provider = RichDecoding.requiredString(object, "provider", allowEmpty: false),
        let scope = RichDecoding.requiredString(object, "scope"),
        scope == "global" || scope == "project"
      else { break }
      let pluginID = RichDecoding.optionalString(object, "pluginId", allowEmpty: false)
      let pluginName = RichDecoding.optionalString(object, "pluginName", allowEmpty: false)
      guard pluginID != .invalid, pluginName != .invalid else { break }
      return .skill(
        name: name,
        path: path,
        invocation: invocation,
        provider: provider,
        scope: scope,
        pluginID: pluginID.value,
        pluginName: pluginName.value
      )
    case "mcp":
      guard let id = RichDecoding.requiredString(object, "id", allowEmpty: false),
        let name = RichDecoding.requiredString(object, "name", allowEmpty: false)
      else { break }
      return .mcp(id: id, name: name)
    default:
      break
    }
    throw RichDomainDecodeError.invalidPendingSteer
  }

  private static func decodePending(_ value: RichJSON) throws -> RichPendingSteer {
    guard let object = value.objectValue,
      let id = RichDecoding.requiredString(object, "id", allowEmpty: false),
      let prompt = RichDecoding.requiredString(object, "prompt"),
      let stagedDecimal = object["stagedAt"]?.decimalValue
    else { throw RichDomainDecodeError.invalidPendingSteer }
    let stagedAt = NSDecimalNumber(decimal: stagedDecimal).doubleValue
    guard stagedAt.isFinite else { throw RichDomainDecodeError.invalidPendingSteer }
    return RichPendingSteer(
      id: id,
      prompt: prompt,
      segments: try decodeOptionalSegments(object),
      stagedAtMilliseconds: stagedAt
    )
  }

  private static func decodeOptionalSegments(
    _ object: [String: RichJSON]
  ) throws -> [RichPromptSegment]? {
    switch RichDecoding.optionalArray(object, "segments") {
    case .absent: return nil
    case .invalid: throw RichDomainDecodeError.invalidPendingSteer
    case .value(let values): return try values.map(decodeSegment)
    }
  }
}

File: android/app/src/main/kotlin/com/poracode/app/chat/RichPendingSteer.kt
package com.poracode.app.chat

import com.poracode.app.model.ClientConnectionId
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject

sealed interface RichPromptSegment {
    data class Text(val content: String) : RichPromptSegment
    data class File(val path: String) : RichPromptSegment
    data class Attachment(val path: String, val mimeType: String? = null) : RichPromptSegment

    data class DiffComment(
        val path: String,
        val lineNumber: Long,
        val side: RichDiffSide,
        val staged: Boolean,
        val body: String,
    ) : RichPromptSegment

    data class Skill(
        val name: String,
        val path: String,
        val invocation: String,
        val provider: String,
        val scope: String,
        val pluginId: String? = null,
        val pluginName: String? = null,
    ) : RichPromptSegment

    data class Mcp(val id: String, val name: String) : RichPromptSegment
}

data class RichPendingSteer(
    val id: String,
    val prompt: String,
    val segments: List<RichPromptSegment>? = null,
    val stagedAtEpochMs: Double,
)

data class RichPendingSteerEnvelope(
    val threadKey: RichThreadKey,
    val pending: RichPendingSteer?,
)

data class RichSetPendingSteerInput(
    val prompt: String,
    val segments: List<RichPromptSegment>? = null,
    val config: JsonObject,
)

object RichPendingSteerDecoder {
    fun decodeEnvelope(
        connectionId: ClientConnectionId,
        value: JsonElement,
    ): RichPendingSteerEnvelope? {
        val objectValue = value.objectOrNull() ?: return null
        if (objectValue.requiredString("type") != "thread-pending-steer") return null
        val threadId = objectValue.requiredString("threadId", allowEmpty = false) ?: return null
        if (!objectValue.containsKey("pending")) return null
        val pending = when (val raw = objectValue["pending"]) {
            null, JsonNull -> null
            else -> decodePending(raw) ?: return null
        }
        return RichPendingSteerEnvelope(RichThreadKey(connectionId, threadId), pending)
    }

    fun decodeSetBody(value: JsonElement): RichSetPendingSteerInput? {
        val objectValue = value.objectOrNull() ?: return null
        val prompt = objectValue.requiredString("prompt", allowEmpty = false) ?: return null
        val config = objectValue["config"]?.objectOrNull() ?: return null
        val segments = decodeOptionalSegments(objectValue) ?: return null
        return RichSetPendingSteerInput(prompt, segments.value, config)
    }

    fun decodeSegment(value: JsonElement): RichPromptSegment? {
        val objectValue = value.objectOrNull() ?: return null
        return when (objectValue.requiredString("kind")) {
            "text" -> objectValue.requiredString("content")?.let(RichPromptSegment::Text)
            "file" -> objectValue.requiredString("path")?.let(RichPromptSegment::File)
            "attachment" -> decodeAttachment(objectValue)
            "diff_comment" -> decodeDiff(objectValue)
            "skill" -> decodeSkill(objectValue)
            "mcp" -> decodeMcp(objectValue)
            else -> null
        }
    }

    private data class OptionalSegments(
        val value: List<RichPromptSegment>?,
    )

    private fun decodePending(value: JsonElement): RichPendingSteer? {
        val objectValue = value.objectOrNull() ?: return null
        val id = objectValue.requiredString("id", allowEmpty = false) ?: return null
        val prompt = objectValue.requiredString("prompt") ?: return null
        val stagedAt = objectValue["stagedAt"]?.finiteDoubleOrNull() ?: return null
        val segments = decodeOptionalSegments(objectValue) ?: return null
        return RichPendingSteer(id, prompt, segments.value, stagedAt)
    }

    private fun decodeOptionalSegments(value: JsonObject): OptionalSegments? {
        return when (val field = value.optionalArray("segments")) {
            RichField.Missing -> OptionalSegments(null)
            RichField.Invalid -> null
            is RichField.Value -> OptionalSegments(decodeSegments(field.value) ?: return null)
        }
    }

    private fun decodeSegments(value: JsonArray): List<RichPromptSegment>? =
        value.map { decodeSegment(it) ?: return null }

    private fun decodeAttachment(value: JsonObject): RichPromptSegment? {
        val path = value.requiredString("path") ?: return null
        val mime = value.optionalString("mimeType")
        if (mime is RichField.Invalid) return null
        return RichPromptSegment.Attachment(path, mime.valueOrNull())
    }

    private fun decodeDiff(value: JsonObject): RichPromptSegment? {
        val path = value.requiredString("path", allowEmpty = false) ?: return null
        val line = value["lineNumber"]?.longOrStrictNull()?.takeIf { it > 0 } ?: return null
        val side = value.requiredString("side")?.let(RichDiffSide::fromWire) ?: return null
        val staged = value["staged"]?.booleanOrStrictNull() ?: return null
        val body = value.requiredString("body", allowEmpty = false) ?: return null
        return RichPromptSegment.DiffComment(path, line, side, staged, body)
    }

    private fun decodeSkill(value: JsonObject): RichPromptSegment? {
        val name = value.requiredString("name", allowEmpty = false) ?: return null
        val path = value.requiredString("path", allowEmpty = false) ?: return null
        val invocation = value.requiredString("invocation", allowEmpty = false) ?: return null
        val provider = value.requiredString("provider", allowEmpty = false) ?: return null
        val scope = value.requiredString("scope")?.takeIf { it == "global" || it == "project" }
            ?: return null
        val pluginId = value.optionalString("pluginId", allowEmpty = false)
        val pluginName = value.optionalString("pluginName", allowEmpty = false)
        if (pluginId is RichField.Invalid || pluginName is RichField.Invalid) return null
        return RichPromptSegment.Skill(
            name,
            path,
            invocation,
            provider,
            scope,
            pluginId.valueOrNull(),
            pluginName.valueOrNull(),
        )
    }

    private fun decodeMcp(value: JsonObject): RichPromptSegment? {
        val id = value.requiredString("id", allowEmpty = false) ?: return null
        val name = value.requiredString("name", allowEmpty = false) ?: return null
        return RichPromptSegment.Mcp(id, name)
    }
}
